specification.md

```diff
@@ -795,6 +795,44 @@ Path? getStderrPath()
 
 Set/get the paths to the standard stream files.
 
+<a name="jobspec-setprelaunch"></a>
+<a name="jobspec-setpreexec"></a>
+<a name="jobspec-setpostexec"></a>
+<a name="jobspec-setpostlaunch"></a>
+```java
+void setPreLaunch(Path script)
+Path? getPreLaunch()
+void setPreExec(Path script)
+Path? getPreExec()
+void setPostExec(Path script)
+Path? getPostExec()
+void setPostLaunch(Path script)
+Path? getPostLaunch()
+```
+
+Set/get the paths to the pre/post exec/launch scripts.
+
+The `PreLaunch` script is sourced on rank 0 of the job before the parallel
+launch of the `executable`.  This script is intended for setting up the
+environment via external systems like `lmod` and `virtualenv` as well as for
+performing actions that should only be done once at the start of a job (e.g.,
+creating a results directory).
+
+The `PreExec` script is sourced by every rank of the job after the parallel
+launch but before the exec of the `executable`. `JPSI_RANK` is set to the rank
+of the current process within the job and accessible by the `PreExec` script.
+This script is intended for setting up any rank-specific environment variables
+and performing actions that should be performed once per rank (e.g., staging in
+files onto local storage).
+
+The `PostExec` script is sourced by every rank of the job after the exit of the
+`executable`.  The `PostExec` script SHALL run if the `executable` fails but it
+may fail to run if the job is killed due to exceeding its walltime.
+
+The `PostLaunch` script is sourced by rank 0 of the job after the exit of the
+`executable` and after the completion of the `PostExec` script on every rank.
+The `PostLaunch` script may fail to run if the job walltime expires, including
+if any of the `PostExec` scripts hang.
```

Comment:
Does `PostLaunch` imply synchronization, i.e., do we wait for *all* ranks to complete, or only for rank 0?  I would suggest the former, to allow the script to act upon global output data, for example.